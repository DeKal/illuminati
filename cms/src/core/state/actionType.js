export const ERROR = 'app/error'
export const APP_TURN_ON_LOADING = 'app/loading/on'
export const APP_TURN_OFF_LOADING = 'app/loading/off'
export const OPEN_TOAST = 'app/toast/open'
export const CLOSE_TOAST = 'app/toast/close'

export const OPEN_DRAWER = 'drawer/openDrawer'
export const CLOSE_DRAWER = 'drawer/closeDrawer'

export const CHANGE_PAGE = 'posts/changePage'
export const CHANGE_ROW_PER_PAGE = 'posts/changeRowPerPage'
export const SET_ORDER = 'posts/setOrder'
export const SET_ORDER_COLUMN = 'posts/setOrderColumn'
export const SET_SELECTED_POST = 'posts/setSelectedPost'
export const SELECT_ALL_POSTS = 'posts/selectAllPosts'
export const CLEAR_SELECTED = 'posts/clearSelected'
export const FETCH_POSTS = 'posts/fetchPosts'

export const FETCH_POST = 'post/fetchPost'
export const EDIT_POST_TITLE = 'post/editPostTitle'
export const EDIT_POST_SLUG = 'post/editPostSlug'
export const EDIT_POST_BRIEF = 'post/editPostBrief'
export const EDIT_POST_AUTHOR = 'post/editPostAuthor'
export const EDIT_POST_TIME = 'post/editPostTime'
export const EDIT_POST_TAGS = 'post/editPostTags'
export const EDIT_POST_CONTENT = 'post/editPostContent'
export const SAVE_EDITED_CONTENT = 'post/saveEditedContent'
export const CREATE_NEW_POST = 'post/createNewPost'
export const INIT_POST = 'post/initPost'
