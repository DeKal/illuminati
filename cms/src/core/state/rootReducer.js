import { combineReducers } from 'redux'

import drawer from 'core/state/drawer/reducers'
import posts from 'posts/state/reducers'

export default combineReducers({
  drawer,
  posts
})
