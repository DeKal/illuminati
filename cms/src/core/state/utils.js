export const toSuccess = actionName => `${actionName}Success`

export const toError = actionName => `${actionName}Error`

export const toRequest = actionName => `${actionName}Request`
