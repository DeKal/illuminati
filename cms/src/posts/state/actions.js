import { createAction } from 'redux-actions'
import {
  CHANGE_PAGE,
  CHANGE_ROW_PER_PAGE,
  SET_ORDER,
  SET_ORDER_COLUMN,
  SET_SELECTED_POST,
  SELECT_ALL_POSTS,
  CLEAR_SELECTED
} from 'core/const/actionType'

export const changePage = createAction(CHANGE_PAGE)

export const changeRowsPerPage = createAction(CHANGE_ROW_PER_PAGE)

export const setOrder = createAction(SET_ORDER)

export const setOrderColumn = createAction(SET_ORDER_COLUMN)

export const setSelectedPost = createAction(SET_SELECTED_POST)

export const selectAllPosts = createAction(SELECT_ALL_POSTS)

export const clearSelected = createAction(CLEAR_SELECTED)
