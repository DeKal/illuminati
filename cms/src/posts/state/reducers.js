import {
  CHANGE_PAGE,
  CHANGE_ROW_PER_PAGE,
  SET_ORDER,
  SET_ORDER_COLUMN,
  SET_SELECTED_POST,
  SELECT_ALL_POSTS,
  CLEAR_SELECTED
} from 'core/const/actionType'

const posts = (state = {}, action) => {
  switch (action.type) {
    case CHANGE_PAGE:
      return {
        ...state,
        page: action.payload
      }
    case CHANGE_ROW_PER_PAGE:
      return {
        ...state,
        page: 0,
        rowsPerPage: action.payload
      }
    case SET_ORDER:
      return {
        ...state,
        order: action.payload
      }
    case SET_ORDER_COLUMN:
      return {
        ...state,
        orderColumn: action.payload
      }
    case SET_SELECTED_POST: {
      const name = action.payload
      const { selectedPosts } = state
      const selectedIndex = state.selectedPosts.indexOf(name)
      let newSelected = []

      if (selectedIndex === -1) {
        newSelected = newSelected.concat(selectedPosts, name)
      } else if (selectedIndex === 0) {
        newSelected = newSelected.concat(selectedPosts.slice(1))
      } else if (selectedIndex === selectedPosts.length - 1) {
        newSelected = newSelected.concat(selectedPosts.slice(0, -1))
      } else if (selectedIndex > 0) {
        newSelected = newSelected.concat(
          selectedPosts.slice(0, selectedIndex),
          selectedPosts.slice(selectedIndex + 1)
        )
      }
      return {
        ...state,
        selectedPosts: newSelected
      }
    }
    case SELECT_ALL_POSTS: {
      const { posts } = state
      const newSelecteds = posts.map(n => n.name)
      return {
        ...state,
        selectedPosts: newSelecteds
      }
    }
    case CLEAR_SELECTED: {
      return {
        ...state,
        selectedPosts: []
      }
    }
    default:
      return state
  }
}

export default posts
