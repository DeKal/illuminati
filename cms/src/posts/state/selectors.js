export const getPage = state => state.posts.page

export const getRowsPerPage = state => state.posts.rowsPerPage

export const getOrder = state => state.posts.order

export const getOrderColumn = state => state.posts.orderColumn

export const getSelectedPosts = state => state.posts.selectedPosts

export const getPosts = state => state.posts.posts
