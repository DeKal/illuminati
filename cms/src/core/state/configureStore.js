import { createStore } from 'redux'
import { persistStore, persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage'
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2'
import rootReducer from 'core/state/rootReducer'

const persistConfig = {
  key: 'root',
  storage,
  stateReconciler: autoMergeLevel2,
  whitelist: ['drawer']
}

const persistedReducer = persistReducer(persistConfig, rootReducer)

const configureStore = initState => {
  const store = createStore(persistedReducer, initState)
  const persistor = persistStore(store)
  return { store, persistor }
}

export default configureStore
