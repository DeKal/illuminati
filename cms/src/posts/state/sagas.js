import { call, put, takeLatest } from 'redux-saga/effects'
import { fetchPostSummary } from 'posts/api/fetch'
import { FETCH_POSTS } from 'core/state/actionType'
import { toSuccess, toError, toRequest } from 'core/state/utils'

function* fetchPosts() {
  try {
    const posts = yield call(fetchPostSummary)
    yield put({ type: toSuccess(FETCH_POSTS), payload: posts })
  } catch (e) {
    yield put({ type: toError(FETCH_POSTS), message: e.message })
  }
}

function* postsSaga() {
  yield takeLatest(toRequest(FETCH_POSTS), fetchPosts)
}

export default postsSaga
