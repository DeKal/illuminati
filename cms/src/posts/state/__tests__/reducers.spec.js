import postsReducer from '../reducers'
import {
  CHANGE_PAGE,
  CHANGE_ROW_PER_PAGE,
  SET_ORDER,
  SET_ORDER_COLUMN,
  SET_SELECTED_POST,
  SELECT_ALL_POSTS,
  CLEAR_SELECTED,
  FETCH_POSTS
} from 'core/state/actionType'
import { toSuccess, toError } from 'core/state/utils'

describe('Posts Reducer ', () => {
  describe('default', () => {
    it('should work corectly', function() {
      const action = {
        type: 'default'
      }
      expect(postsReducer({}, action)).toMatchSnapshot()
      expect(() => {
        postsReducer()
      }).toThrow(TypeError)
    })
  })
  describe('CHANGE_PAGE', () => {
    it('should work corectly', function() {
      const action = {
        type: CHANGE_PAGE,
        payload: 1
      }
      expect(postsReducer({}, action)).toMatchSnapshot()
    })
  })
  describe('CHANGE_ROW_PER_PAGE', () => {
    it('should work corectly', function() {
      const action = {
        type: CHANGE_ROW_PER_PAGE,
        payload: 10
      }
      expect(postsReducer({}, action)).toMatchSnapshot()
    })
  })
  describe('SET_ORDER', () => {
    it('should work corectly', function() {
      const action = {
        type: SET_ORDER,
        payload: 'asc'
      }
      expect(postsReducer({}, action)).toMatchSnapshot()
    })
  })
  describe('SET_ORDER_COLUMN', () => {
    it('should work corectly', function() {
      const action = {
        type: SET_ORDER_COLUMN,
        payload: 'author'
      }
      expect(postsReducer({}, action)).toMatchSnapshot()
    })
  })

  describe('SELECT_ALL_POSTS', () => {
    it('should work corectly', function() {
      const posts = [{ slug: 'a' }, { slug: 'b' }]
      const action = {
        type: SELECT_ALL_POSTS
      }
      expect(postsReducer({ posts }, action)).toMatchSnapshot()
    })
  })

  describe('CLEAR_SELECTED', () => {
    it('should work corectly', function() {
      const action = {
        type: CLEAR_SELECTED
      }
      const selectedPosts = [{ slug: 'a' }, { slug: 'b' }]
      expect(postsReducer({ selectedPosts }, action)).toMatchSnapshot()
    })
  })

  describe('FETCH_POSTS', () => {
    it('success', function() {
      const action = {
        type: toSuccess(FETCH_POSTS),
        payload: 'success payload'
      }
      expect(postsReducer({}, action)).toMatchSnapshot()
    })

    it('error', function() {
      const action = {
        type: toError(FETCH_POSTS),
        payload: 'faild payload'
      }
      expect(postsReducer({}, action)).toMatchSnapshot()
    })
  })

  describe('SET_SELECTED_POST', () => {
    it('add new selected post', function() {
      const selectedPosts = ['a', 'b', 'c', 'd']
      const action = {
        type: SET_SELECTED_POST,
        payload: 'e'
      }

      expect(postsReducer({ selectedPosts }, action)).toMatchSnapshot()
    })

    it('remove a selected post at the beginning', function() {
      const selectedPosts = ['a', 'b', 'c', 'd']
      const action = {
        type: SET_SELECTED_POST,
        payload: 'a'
      }

      expect(postsReducer({ selectedPosts }, action)).toMatchSnapshot()
    })

    it('remove a selected post at the middle', function() {
      const selectedPosts = ['a', 'b', 'c', 'd']
      const action = {
        type: SET_SELECTED_POST,
        payload: 'b'
      }

      expect(postsReducer({ selectedPosts }, action)).toMatchSnapshot()
    })

    it('remove a selected post at the end', function() {
      const selectedPosts = ['a', 'b', 'c', 'd']
      const action = {
        type: SET_SELECTED_POST,
        payload: 'd'
      }

      expect(postsReducer({ selectedPosts }, action)).toMatchSnapshot()
    })
  })
})
