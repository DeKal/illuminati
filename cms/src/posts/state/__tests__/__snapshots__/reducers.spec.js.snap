// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts Reducer  CHANGE_PAGE should work corectly 1`] = `
Object {
  "page": 1,
}
`;

exports[`Posts Reducer  CHANGE_ROW_PER_PAGE should work corectly 1`] = `
Object {
  "page": 0,
  "rowsPerPage": 10,
}
`;

exports[`Posts Reducer  CLEAR_SELECTED should work corectly 1`] = `
Object {
  "selectedPosts": Array [],
}
`;

exports[`Posts Reducer  FETCH_POSTS error 1`] = `
Object {
  "posts": Array [],
}
`;

exports[`Posts Reducer  FETCH_POSTS success 1`] = `
Object {
  "posts": "success payload",
}
`;

exports[`Posts Reducer  SELECT_ALL_POSTS should work corectly 1`] = `
Object {
  "posts": Array [
    Object {
      "slug": "a",
    },
    Object {
      "slug": "b",
    },
  ],
  "selectedPosts": Array [
    "a",
    "b",
  ],
}
`;

exports[`Posts Reducer  SET_ORDER should work corectly 1`] = `
Object {
  "order": "asc",
}
`;

exports[`Posts Reducer  SET_ORDER_COLUMN should work corectly 1`] = `
Object {
  "orderColumn": "author",
}
`;

exports[`Posts Reducer  SET_SELECTED_POST add new selected post 1`] = `
Object {
  "selectedPosts": Array [
    "a",
    "b",
    "c",
    "d",
    "e",
  ],
}
`;

exports[`Posts Reducer  SET_SELECTED_POST remove a selected post at the beginning 1`] = `
Object {
  "selectedPosts": Array [
    "b",
    "c",
    "d",
  ],
}
`;

exports[`Posts Reducer  SET_SELECTED_POST remove a selected post at the end 1`] = `
Object {
  "selectedPosts": Array [
    "a",
    "b",
    "c",
  ],
}
`;

exports[`Posts Reducer  SET_SELECTED_POST remove a selected post at the middle 1`] = `
Object {
  "selectedPosts": Array [
    "a",
    "c",
    "d",
  ],
}
`;

exports[`Posts Reducer  default should work corectly 1`] = `Object {}`;
